8254877: GCLogPrecious::_lock rank constrains what locks you are allowed to have when crashing

stefank opened the pull request:
This is an alternative version of the fix proposed in 900:
https://github.com/openjdk/jdk/pull/900

Erik's description:
> Today, when you crash, the GCLogPrecious::_lock is taken. This effectively limits you to only get clean crash reports if you crash or assert without holding a lock of rank tty or lower. It is arguably difficult to know what locks you are going to have when crashing. Therefore, I don't think the precious GC log should constrain possible crashing contexts in that fashion.

As Erik mentioned in that PR, I'd like to retain the ability to easily dump the precious log when debugging. The proposed fix changes the Mutex to a Semaphore, and use trywait to safely access the buffer. In the unlikely event that another thread is holding the lock, the hs_err printer skips printing the log.

This also makes it possible to call precious logging from within the stack watermark processing code. I think there's a possibility that we might call the following error logging, when we fail to commit memory for a ZPage, when relocating, during stack watermark processing:
`log_error_p(gc)("Failed to commit memory (%s)", err.to_string());`

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux additional | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (8/8 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |  ✔️ (9/9 passed) | ❌ (2/9 failed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

**Failed test tasks**
- [Linux x86 (hs/tier1 compiler)](https://github.com/stefank/jdk/runs/1458351316)
- [Linux x86 (jdk/tier1 part 3)](https://github.com/stefank/jdk/runs/1458351197)

### Issue
 * [JDK-8254877](https://bugs.openjdk.java.net/browse/JDK-8254877): GCLogPrecious::_lock rank constrains what locks you are allowed to have when crashing


### Reviewers
 * [Erik Österlund](https://openjdk.java.net/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to 9888692b70e35e4a284b37be9c4c231d7056ce7d


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/903/head:pull/903`
`$ git checkout pull/903`

pliden commented on src/hotspot/share/gc/shared/gcLogPrecious.cpp:
```diff
@@ -77,11 +75,20 @@ void GCLogPrecious::vwrite_and_debug(LogTargetHandle log,
+    st->print_cr("%s", _lines->base());
+  }
+
+  _lock->signal();
```
As we discussed off-line, perhaps we might want to print something even if the log isn't initialized and/or is empty. Something like:
```
  st->print_cr("GC Precious Log:");

  if (_lines == NULL) {
    st->print_cr("<Not initialized>");
    return;
  }

  if (!_lock->trywait()) {
    st->print_cr("<Skipped>");
    return;
  }

  if (_lines->size() == 0) {
    st->print_cr("<Empty>");
  } else {
    st->print_cr("%s", _lines->base());
  }

  _lock->signal();
```
You decide. Looks good otherwise.

stefank replied:
Updated with suggestion. I also added extra newlines to make the output look pretty. That was needed because _lines->base() is always terminated with a newline.

fisk reviewed (approved):
Looks good.

albertnetymk commented on src/hotspot/share/gc/shared/gcLogPrecious.cpp:
```diff
@@ -23,20 +23,16 @@
 stringStream* GCLogPrecious::_lines = NULL;
 stringStream* GCLogPrecious::_temp = NULL;
-Mutex* GCLogPrecious::_lock = NULL;
+Semaphore* GCLogPrecious::_lock = NULL;
```
Maybe renaming `_lock` to `_semaphore`? Additionally, since it's a binary semaphore, `new Semaphore(1)`, some comments explaining why `Mutex` is **not** suitable could avoid some future confusions.

PS: not a review, just a comment in passing.

stefank replied:
It's used as a lock, so I think the name `_lock` is appropriate. Instead I introduced a new class: `SemaphoreLock`, to make the code more readable (IMHO). Also added a comment. Hopefully, this addressed your comments.

albertnetymk replied:
Thank you; it does look more readable. BTW, mutex has `try_lock` as well.

stefank replied:
I see. Then I'll remove that part of the new comment.

stefank commented:
In the latest update I added two new helper classes: `SemaphoreLock` and `SemaphoreLocker`. I think this makes the code nicer. Since those classes are more broadly used, I'll go a head and split them out into a separate PR.

stefank commented:
Forked off the SemaphoreLock part into https://github.com/openjdk/jdk/pull/927

stefank commented:
/label add hotspot

pliden commented:
Updates look good!

stefank commented:
I've thrown away the Semaphore implementation and replaced it with Patricio's new try_lock_without_range_check function. I've also changed the lock rank to be as low as possible. We'll still get a lock rank reordering problem if we crash while holding this lock, because EventLog takes its lock and it is of the same rank. I intend to address that with JDK-8256382.

pliden reviewed (commented):
Update looks good. Just one question, is there a reason you need/want to change the rank from tty to event, even though we call try_lock_without_rank_check() in the sensitive part?

fisk reviewed (approved):
Looks good. Quite entertaining that the "leaf" rank is so far away from being a leaf level rank nowadays.

stefank commented:
> Looks good. Quite entertaining that the "leaf" rank is so far away from being a leaf level rank nowadays.

I had the exact same thought. :(

Thanks for reviewing @pliden and @fisk

stefank commented:
/integrate
